File: src/csharp/TLE.cs
using System;
using System.Globalization;

namespace sgp4
{


/**
 * This class knows how to parse two line element sets and access the individual elements.
 * 
 * @author aholinch
 *
 */
public class TLE 
{
    protected ElsetRec rec = null;
    
    protected String line1 = null;
    protected String line2 = null;
    
    protected String intlid = null;
    protected String objectID = null;
    protected DateTime epoch;
    protected double ndot = 0;
    protected double nddot = 0;
    protected double bstar = 0;
    protected int elnum = 0;
    protected double incDeg = 0;
    protected double raanDeg = 0;
    protected double ecc = 0;
    protected double argpDeg = 0;
    protected double maDeg = 0;
    protected double n = 0;
    protected int revnum = 0;
    
    protected String parseErrors = null;
    
    protected int sgp4Error = 0;
    
    public TLE()
    {
        
    }
    
    public TLE(String line1, String line2)
    {
        parseLines(line1,line2);
    }

    public double[][] getRV(DateTime d)
    {
        double t = d.Ticks;
        t -= epoch.Ticks;
        t/=(60000.0*10000.0);
        
        return getRV(t);
    }

    public double[][] getRV(double minutesAfterEpoch)
    {
        double []r = new double[3];
        double []v = new double[3];
        
        rec.error = 0;
        SGP4.sgp4(rec, minutesAfterEpoch, r, v);
        sgp4Error = rec.error;
        double [][] outv = new double[2][];
        outv[0]=r;
        outv[1]=v;
        return outv;
    }
    
    public int getSgp4Error()
    {
        return sgp4Error;
    }
    
    public ElsetRec getElsetRec()
    {
        return rec;
    }
    
    public void setElsetRec(ElsetRec er)
    {
        rec = er;
    }
    
    public String getParseErrors()
    {
        return parseErrors;
    }
    
    public String getIntlID()
    {
        return intlid;
    }
    
    public void setIntlID(String id)
    {
        intlid = id;
    }
    
    public String getObjectID()
    {
        return objectID;
    }
    
    public void setObjectID(String id)
    {
        objectID = id;
    }
    
    public DateTime getEpoch()
    {
        return epoch;
    }
    
    public void setEpoch(DateTime d)
    {
        epoch = d;
    }
    
    public double getNDot()
    {
        return ndot;
    }
    
    public void setNDot(double val)
    {
        ndot = val;
    }
    
    public double getNDDot()
    {
        return nddot;
    }
    
    public void setNDDot(double val)
    {
        nddot = val;
    }
    
    public double getBstar()
    {
        return bstar;
    }
    
    public void setBstar(double val)
    {
        bstar = val;
    }
    
    public int getElNum()
    {
        return elnum;
    }
    
    public void setElNum(int num)
    {
        elnum = num;
    }
    
    public double getIncDeg()
    {
        return incDeg;
    }
    
    public void setIncDeg(double val)
    {
        incDeg = val;
    }
    
    public double getRaanDeg()
    {
        return raanDeg;
    }
    
    public void setRaanDeg(double val)
    {
        raanDeg = val;
    }
    
    public double getEcc()
    {
        return ecc;
    }
    
    public void setEcc(double val)
    {
        ecc = val;
    }
    
    public double getArgpDeg()
    {
        return argpDeg;
    }
    
    public void setArgpDeg(double val)
    {
        argpDeg = val;
    }
    
    public double getMaDeg()
    {
        return maDeg;
    }
    
    public void setMaDeg(double val)
    {
        maDeg = val;
    }
    
    public double getN()
    {
        return n;
    }
    
    public void setN(double val)
    {
        n = val;
    }
    
    public int getRevNum()
    {
        return revnum;
    }
    
    public void setRevNum(int val)
    {
        revnum = val;
    }
    
    public String getLine1()
    {
        return line1;
    }
    
    public String getLine2()
    {
        return line2;
    }
    
    /**
     * Parses the two lines optimistically.  No exceptions are thrown but some parse errors will
     * be accumulated as a string.  Call getParseErrors() to see if there are any.
     * 
     * @param line1
     * @param line2
     */
    public void parseLines(String line1, String line2)
    {
        parseErrors = null;
        rec = new ElsetRec();

        this.line1 = line1;
        this.line2 = line2;
        
        if(line1 == null || line1.Trim().Length<68) // we can live without checksum
        {
            addParseError("line1 too short");
        }
        
        if(line2 == null || line2.Trim().Length<68) // we can live without checksum
        {
            addParseError("line2 too short");
        }
        
        if(parseErrors != null) return;
        

        
        //objectNum = (int)gd(line1,2,7);
        //if(objectNum != (int)gd(line2,2,7))addParseError("ids don't match");
        objectID = line1.Substring(2,5).Trim();
        
        rec.classification = line1[7];
           //          1         2         3         4         5         6
           //0123456789012345678901234567890123456789012345678901234567890123456789
    //line1="1 00005U 58002B   00179.78495062  .00000023  00000-0  28098-4 0  4753";
    //line2="2 00005  34.2682 348.7242 1859667 331.7664  19.3264 10.82419157413667";

        intlid = line1.Substring(9, 17-9).Trim();
        epoch = parseEpoch(line1.Substring(18,32-18).Trim());
        ndot = gdi(line1[33],line1,35,44);
        nddot = gdi(line1[44],line1,45,50);
        nddot *= Math.Pow(10.0, gd(line1,50,52));
        bstar = gdi(line1[53],line1,54,59);
        bstar *= Math.Pow(10.0d, gd(line1,59,61));
        
        elnum = (int)gd(line1,64,68);
        
        incDeg = gd(line2,8,16);
        raanDeg = gd(line2,17,25);
        ecc = gdi('+',line2,26,33);
        argpDeg = gd(line2,34,42);
        maDeg = gd(line2,43,51);
        
        n = gd(line2,52,63);
        
        revnum = (int)gd(line2,63,68);
        
        setValsToRec();
    }
    
    /**
     * Set the values to the ElsetRec object.
     */
    protected void setValsToRec()
    {
        double deg2rad = Math.PI / 180.0;          //   0.0174532925199433
        double xpdotp = 1440.0 / (2.0 * Math.PI);  // 229.1831180523293

        rec.elnum = elnum;
        rec.revnum = revnum;
        rec.satid = objectID;
        rec.bstar = bstar;
        rec.inclo = incDeg*deg2rad;
        rec.nodeo = raanDeg*deg2rad;
        rec.argpo = argpDeg*deg2rad;
        rec.mo = maDeg*deg2rad;
        rec.ecco = ecc;
        rec.no_kozai = n/xpdotp;
        rec.ndot = ndot / (xpdotp*1440.0d);
        rec.nddot = nddot / (xpdotp*1440.0d*1440.0d);
        
        SGP4.sgp4init('a', rec);
    }
    
    /**
     * Parse the tle epoch format to a date.  
     * 
     * @param str
     * @return
     */
    protected DateTime parseEpoch(String str)
    {
        int year = (int)Double.Parse(str.Substring(0,2).Trim());
        
        rec.epochyr=year;
        if(year > 56)
        {
            year += 1900;
        }
        else
        {
            year += 2000;
        }
        
        int doy = (int)Double.Parse(str.Substring(2,5-2).Trim());
        double dfrac = Double.Parse("0"+str.Substring(5).Trim(),CultureInfo.InvariantCulture);
        
        rec.epochdays = doy;
        rec.epochdays += dfrac;
        
        
        dfrac *= 24.0d;
        int hr = (int)dfrac;
        dfrac = 60.0d*(dfrac - hr);
        int mn = (int)dfrac;
        dfrac = 60.0d*(dfrac - mn);
        int sc = (int)dfrac;
        

        
        dfrac = 1000.0d*(dfrac-sc);
        int milli = (int)dfrac;
        
        /* 
        gc.set(Calendar.YEAR, year);
        gc.set(Calendar.DAY_OF_YEAR, doy);
        gc.set(Calendar.HOUR_OF_DAY, hr);
        gc.set(Calendar.MINUTE, mn);
        gc.set(Calendar.SECOND, sc);
        gc.set(Calendar.MILLISECOND, milli);
        */
        DateTime mepoch = new DateTime(year,1,1);
        mepoch = mepoch.AddDays(doy-1);
        mepoch = mepoch.AddHours(hr);
        mepoch = mepoch.AddMinutes(mn);
        mepoch = mepoch.AddSeconds(sc);
        mepoch = mepoch.AddMilliseconds(dfrac);

        // Specify that this is UTC
        mepoch = DateTime.SpecifyKind(mepoch,DateTimeKind.Utc);

        double sec = ((double)sc)+dfrac/1000.0d;
        int mon = mepoch.Month;
        int day = mepoch.Day;
        double []jd = SGP4.jday(year, mon, day, hr, mn, sec);
        rec.jdsatepoch = jd[0];
        rec.jdsatepochF = jd[1];


        return mepoch;
    }
    
    protected void addParseError(String err)
    {
        if(parseErrors == null || parseErrors.Trim().Length == 0)
        {
            parseErrors = err;
        }
        else
        {
            parseErrors = parseErrors+"; "+err;
        }
    }
    
    /**
     * parse a double from the substring.
     * 
     * @param str
     * @param start
     * @param end
     * @return
     */
    protected double gd(String str, int start, int end)
    {
        return gd(str,start,end,0);
    }
    
    /**
     * parse a double from the substring.
     * 
     * @param str
     * @param start
     * @param end
     * @param defVal
     * @return
     */
    protected double gd(String str, int start, int end, double defVal)
    {
        double num = defVal;
        try{num = Double.Parse(str.Substring(start,end-start).Trim(),CultureInfo.InvariantCulture);}catch(Exception ex){}
        return num;
    }
    
    /**
     * parse a double from the substring after adding an implied decimal point.
     * 
     * @param str
     * @param start
     * @param end
     * @return
     */
    protected double gdi(char sign, String str, int start, int end)
    {
        return gdi(sign,str,start,end,0);
    }
    
    /**
     * parse a double from the substring after adding an implied decimal point.
     * 
     * @param str
     * @param start
     * @param end
     * @param defVal
     * @return
     */
    protected double gdi(char sign, String str, int start, int end, double defVal)
    {
        double num = defVal;
        try{num = Double.Parse("0."+str.Substring(start,end-start).Trim(),CultureInfo.InvariantCulture);}catch(Exception ex){}
        if(sign == '-') num *= -1.0d;
        return num;
    }

}

}

File: src/mat/TLE.m
%% Two line element set
%% Author: aholinch
classdef TLE < handle
  properties

    rec = 0;
    
    line1 = '';
    line2 = '';
    
    intlid = '';
    objectID = '';
    epoch = 0;
    ndot = 0;
    nddot = 0;
    bstar = 0;
    elnum = 0;
    incDeg = 0;
    raanDeg = 0;
    ecc = 0;
    argpDeg = 0;
    maDeg = 0;
    n = 0;
    revnum = 0;
    
    sgp4Error = 0;
  % end properties
  end

  methods    
    function obj = TLE(l1, l2)
        parseLines(obj,l1,l2);
    end

    function parseLines(this, l1, l2)
    	this.rec = elsetrec();
    	this.line1 = l1;
    	this.line2 = l2;
        line1 = l1;
        line2 = l2;

        this.objectID = strtrim(line1(3:7));
        while(this.objectID(1)=="0")
            this.objectID = this.objectID(2:length(this.objectID));
        end

        this.intlid = strtrim(line1(10:18));
        tmp = line1(34:43);
        tmp(1) = '0';
        this.ndot = str2double(tmp);

        tm = 1;
        if strcmp('-',line1(45:45))
            tm = -1;
        end
        tmp = line1(44:50);
        tmp(1) ='0';
        tmp(2) ='.';
        this.nddot = str2double(tmp);
        ev = str2double(line1(51:52));
        this.nddot = this.nddot * (10^ev);
        this.nddot = tm*this.nddot;

        tm = 1;
        if strcmp('-',line1(54:54))
            tm = -1;
        end
        tmp = line1(53:59);
        tmp(1) ='0';
        tmp(2) ='.';
        this.bstar = str2double(tmp);
        ev = str2double(line1(60:61));
        this.bstar = this.bstar * (10^ev);
        this.bstar = tm*this.bstar;

        this.elnum = str2double(strtrim(line1(65:68)));
%          1         2         3         4         5         6
%0123456789012345678901234567890123456789012345678901234567890123456789
% 1 00005U 58002B   00179.78495062  .00000023  00000-0  28098-4 0  4753";
% 2 00005  34.2682 348.7242 1859667 331.7664  19.3264 10.82419157413667";
        this.epoch = parseEpoch(this,line1(19:32));
        this.incDeg = str2double(strtrim(line2(9:16)));
        this.raanDeg = str2double(strtrim(line2(18:25)));
        tmp = line2(25:33);
        tmp(1)='0';
        tmp(2)='.';
        this.ecc = str2double(strtrim(tmp));
        this.argpDeg = str2double(strtrim(line2(35:42)));
        this.maDeg = str2double(strtrim(line2(44:51)));
        this.n = str2double(strtrim(line2(53:63)));
        this.revnum = str2double(strtrim(line2(64:68)));

        this.rec.classification = line1(8);
        setValsToRec(this)
    end

    function setValsToRec(this)
        deg2rad = pi / 180.0;          %   0.0174532925199433
        xpdotp = 1440.0 / (2.0 * pi);  % 229.1831180523293

        this.rec.elnum = this.elnum;
        this.rec.revnum = this.revnum;
        this.rec.satid = this.objectID;
        this.rec.bstar = this.bstar;
        this.rec.inclo = this.incDeg*deg2rad;
        this.rec.nodeo = this.raanDeg*deg2rad;
        this.rec.argpo = this.argpDeg*deg2rad;
        this.rec.mo = this.maDeg*deg2rad;
        this.rec.ecco = this.ecc;
        this.rec.no_kozai = this.n/xpdotp;
        this.rec.ndot = this.ndot / (xpdotp*1440.0);
        this.rec.nddot = this.nddot / (xpdotp*1440.0*1440.0);
        SGP4.sgp4init('a', this.rec);
    end

    function ep = parseEpoch(this,str)
        year = str2double(strtrim(str(1:2)));
        if(year > 56)
            year = year + 1900;
        else
            year = year + 2000;
        end 

        doy = str2double(strtrim(str(3:5)));
        str(5) = '0';
        dfrac = str2double(strtrim(str(5:14)));
        dfrac = 24.0*dfrac;
        hr = floor(dfrac);
        dfrac = 60.0*(dfrac-hr);
        min = floor(dfrac);
        dfrac = 60.0*(dfrac-min);
        sec = dfrac;
        sc = floor(dfrac);
        milli = 1000.0*(dfrac-sc);

        mons = [31 28 31 30 31 30 31 31 30 31 30 31];

        % some distros don't have this function
        %if(is_leap_year(year))
        il = ((rem (year, 4) == 0 & rem (year, 100) ~= 0) | rem (year, 400) == 0);
        if(il)
            mons(2) = 29;
        end

        mon = 1;
        days = doy;
        while( mon < 12 && days > mons(mon))
            days = days - mons(mon);
            mon = mon + 1;
        end

        jda = SGP4.jday(year,mon,days,hr,min,sec);
        this.rec.jdsatepoch = jda(1);
        this.rec.jdsatepochF = jda(2);
        ep = 0;
    end

    function rv = getRV(this,minutesAfterEpoch)
    	this.rec.error = 0;
        rvhand = rvhandle();
    	SGP4.sgp4(this.rec, minutesAfterEpoch, rvhand);
    	this.sgp4Error = this.rec.error;
        r = rvhand.r;
        v = rvhand.v;
        rv = [r v];
    end
  % end methods
  end

% end class
end
